### DISCUSSION TITLE
Access Violation in .NET accessing the solution value of SlackVar or TransitVar in Routing

### DISCUSSION THREAD

**Discussion by anghy23** [google/or-tools#3142] [Routing (and legacy CP) questions] [answered]
**What version of OR-Tools and what language are you using?**
Version: 9.2
Language: C#

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
Routing Solver

**What operating system (Linux, Windows, ...) and version?**
Windows

**What did you do?**
1. Use the example VrpTimeWindows.cs
2. In PrintSolution add the lines
   var timeSlackVar = timeDimension.SlackVar(index);
   Console.WriteLine("idx:{0} value:{1} -> ", manager.IndexToNode(index), solution.Value(timeSlackVar));

**What did you expect to see**
The value of the slack variable associated to the Time dimension

**What did you see instead?**
Access Violation Exception

**Comment by lperron** [comment, marked as answer]
most likely a wrong index (-1). How was it computed.

**Reply by anghy23** [reply]
Same index used with timeDimension.CumulVar(index), as it is computed in the example.
With the variable returned by CumulVar it works, with the variables returned by SlackVar and TransitVar it gives AccessViolation.
The Access Violation is given by solution.Value(timeSlackVar), not by timeDimension.SlackVar(index)

**Reply by Mizux** [reply]
By default SlackVar are **not** exported in the solution/assignment object, you need to add it for each node (except end node which obviously don't have slack):

example in python:
https://github.com/google/or-tools/blob/dd28bdf6e56f1d71dfc33a77657a99c69dc32efb/ortools/constraint_solver/samples/cvrptw.py#L197-L199

**Reply by anghy23** [reply]
Yes, it worked, thank you.

#761 was useful to understand the meaning of Min and Max of CumulVar and SlackVar, but I cannot understand why TransitVar should have a range, that sometimes doesn't include the real transit value in the distance matrix.

**Comment by lperron** [comment]
You need to add them to the solution template. There should be some code for that.
